src/manual/devel/best-practices/microservices.md

```diff
@@ -0,0 +1,307 @@
+// If this file is executed launch component from first argument
+if (!module.parent) {
+  load(process.argv[2], {
+    process: process.argv[2],
+    profile: process.env.NODE_ENV,
+  }).catch(err => {
+    console.log(err.stack);
+    process.exit(1);
+  });
+}
+
+// Export load for tests
+module.exports = load;
+```
+
+Entries in `Procfile`, then, look like `web: node lib/main.js server`.
+The web service should always be the component named `server`.
+All services, including those run from the Heroku scheduler, should start like this, via `lib/main.js`.
+
+### azure-entities
+
+Each Azure table should be defined in `src/data.js` using a cascade of `configure` calls, one for each version:
+
+```js
+var MyEntity = Entity.configure({
+  version:              1,
+}).configure({
+  version:              2,
+  migrate: function(item) {
+    // ...
+  },
+}).configure({
+  version:              3,
+  migrate: function(item) {
+    // ...
+  },
+});
+```
+
+The result is a `MyEntity` class that can be setup with additional configuration in a loader component, in `src/main.js`:
+
+```js
+{
+  MyEntity: {
+    requires: ['cfg', 'process', 'monitor'],
+    setup: ({cfg, process, monitor}) => {
+      return data.MyEntity.setup(_.defaults({
+        table:        cfg.app.hookTable,
+        monitor:      monitor.prefix(cfg.app.hookTable.toLowerCase()),
+        component:    cfg.app.component,
+        process,
+      }, cfg.azureTable, cfg.taskcluster));
+    },
+  },
+}
+```
+
+In `src/data.js`, it is common to add utility methods for the entity type.
+For entities which can be fetched via an API, a `json` method is common:
+
+```js
+MyEntity.prototype.json = () => {
+  return {
+    foo: this.foo,
+  };
+};
+```
+
+### taskcluster-lib-api
+
+The API definition should be in `src/v1.js`:
+
+```js
+var api = new API({
+  // ...
+});
+
+// Export api
+module.exports = api;
+
+/** Get hook groups **/
+api.declare({
+  // ...
+});
+// ...
+```
+
+This is then imported and set up in `src/main.js`:
+
+```js
+{
+  router: {
+    requires: ['cfg', 'profile', 'validator', 'monitor'],
+    setup: ({cfg, profile, validator, monitor}) => {
+      return v1.setup({
+        context: {},
+        authBaseUrl:      cfg.taskcluster.authBaseUrl,
+        publish:          profile === 'production',
+        baseUrl:          cfg.server.publicUrl + '/v1',
+        referencePrefix:  'myservice/v1/api.json',
+        aws:              cfg.aws,
+        validator,
+        monitor,
+      });
+    },
+  },
+}
+```
+
+#### Error Handling
+
+Do not use `res.status(..)` to return error messages.
+Instead, use `res.reportError(code, message, details)`.
+The library provides a few codes, but it is easy to add new codes in the API constructor:
+
+```js
+var api = new API({
+  description: [
+    // ...
+    '',
+    '## Error Codes',
+    '',
+    '* `SomethingReallyBad` (472) - you're really not going to like this',
+  ].join('\n'),
+  errorCodes: {
+    SomethingReallyBad: 472,
```

Comment:
Most of them say "Only for..", except InvalidInput, ResourceNotFound, and ResourceConflict.  Is it bad to make more codes that map to 400?